### PR TITLE
Add DiceCEEdge loss recipe

### PR DESCRIPTION
**Pull request by lkdci**
## DiceEdge recipe results
New recipe for STDC training, using Edge attention loss and Dice loss, replacing the not stable detail loss.
Results:
| model | mIoU | previous %mIoU diff |
| ------ | ----- | --------------- |
| STDC1Seg50 | 75.11 | 0.75 |
| STDC2Seg50 | 76.44 | 1.17 |

**This PR includes:**
## Target / one-hot to binary edge map util functions:
create edge feature maps function from one hot tensor. using dilation erosion with convolution.
* Edge widths can be adjusted with kernel_size argument, typical edge width is `kernel_size -1`
* input can be a multi-class or one class one hot tensor with shape [B, C, H, W]
* In case of multi-class the result tensor can be flattened to one channel edge map, or keep the channel dimension so each channel is the edge features per class. see below example.
* Only odd kernel_size are valid to prevent dimension change or pixel shifting.

Cityscapes example:
* kernel_size: left k=3, right k=9.

kernel_city

* Flatten_channels = False: left: person class edges, right: car class edges

cls_city

## Mask Loss (edge loss):

Mask attention loss, for region enforced loss functions i.e edge loss.
* support many criterion functions, such as, CrossEntropyLoss, BCEWithLogitsLoss, MSELoss, SL1Loss.
* Losses is composed of regular loss and mask loss, which are weighted with the `loss_weights` argument.
* unit tests added.

## DiceCEEdgeLoss main class and recipe:
* DiceCEEdgeLoss: main loss class for losses combination of dice, edge-ce and auxiliaries.
* yaml only recipe

### CODE REVIEW

**Review by shaydeci — CHANGES_REQUESTED** [review]
Inline

File: src/super_gradients/training/losses/mask_loss.py [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+        if loss_weights[1] <= 0:
+            raise ValueError("If no loss weight is applied on mask samples, consider using simply criterion")
+
+        self.criterion = criterion
+        self.loss_weights = loss_weights
+
+    def forward(self, predict: torch.Tensor, target: torch.Tensor, mask: torch.Tensor):
+        criterion_loss = self.criterion(predict, target)
+
+        mask = self._broadcast_mask(mask, criterion_loss.size())
```
[the remark this answers is not in the document]
Comment (lkdci) [reply src/super_gradients/training/losses/mask_loss.py:43]: This class support different criterion function, the returned cost tensor is in different format, for CE the returned loss shape is same as the target tensor [B, H, W], for BCE / L1 / MSE the shape is same as the predictions [B, C, H, W].
On the other hand the mask can be either 4dim tensor or 3dim.
An example use case is using  Binary CE with multiple classes, prediction and target (usually one-hot) are with shape [BxCxHxW], while the mask can be with different shapes, [Bx1xHxW] | [BxCxHxW] this method should broadcast the mask tensor in the channel dimension.

File: src/super_gradients/training/losses/mask_loss.py [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+        mask_loss = criterion_loss * mask
+
+        if self.reduction == LossReduction.NONE.value:
+            return criterion_loss * self.loss_weights[0] + mask_loss * self.loss_weights[1]
```
[the remark this answers is not in the document]
Comment (lkdci) [reply src/super_gradients/training/losses/mask_loss.py:47]: Not all loss functions natively support spatial weights.

File: src/super_gradients/recipes/cityscapes_stdc_seg50_dice_edge.yaml [hunk 3]
```diff
@@ -0,0 +1,43 @@
+#  STDC segmentation training example with Cityscapes dataset.
```
[the remark this answers is not in the document]
Comment (lkdci) [reply src/super_gradients/recipes/cityscapes_stdc_seg50_dice_edge.yaml]: Currently this PR improve the results of Seg50 models. Thus there is a recipe file for seg50 models `cityscapes_stdc_seg50`, recipe for seg75 which wasn't changed here `cityscapes_stdc_seg75`
Updated pretrained checkpoints for STDCX-seg50 models

File: src/super_gradients/training/utils/segmentation_utils.py [hunk 4]
```diff
@@ -135,3 +135,59 @@ def visualize_batches(dataloader, module, visualization_path, num_batches=1, und
...
+    :return: one_hot edge torch.Tensor.
+    """
+    if kernel_size < 0 or kernel_size % 2 == 0:
+        raise ValueError(f"kernel size must be an odd positive values, such as [1, 3, 5, ..], found: {kernel_size}")
+    _kernel = torch.ones(x.size(1), 1, kernel_size, kernel_size, dtype=torch.float32, device=x.device)
+    padding = (kernel_size - 1) // 2
+    # Use replicate padding to prevent class shifting and edge formation at the image boundaries.
+    padded_x = F.pad(x.float(), mode="replicate", pad=[padding] * 4)
+
+    dilation = torch.clamp(
```
[the remark this answers is not in the document]
Comment (lkdci) [reply src/super_gradients/training/utils/segmentation_utils.py:160]: We considered different way to transform a label tensor into edges feature map, such as, Laplacian filter and scipy euclidean distance transform. Dilation -Erosion using grouped convolution worked the best in term of speed and devices compitability, while preserving the edges. You are right that when picking a large kernel_size the edges will be coarse and pixels might merged if their distance is lower than `kernel_size -1`

**Comment by lkdci** [comment]
Addressed, the results improved in this PR are STDC-Seg50 models.

**Comment by lkdci** [comment]
Hi @shaydeci , would really appreciate your re-review on this one :)

File: src/super_gradients/training/utils/segmentation_utils.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (shaydeci) [reply src/super_gradients/training/utils/segmentation_utils.py:160]: Not questioning what worked best (: . 
My comment's intent was to get more readable, clean code by extracting all of the edge calculation to a well documented method.

[the remark this answers is not in the document]
Comment (lkdci) [reply src/super_gradients/training/utils/segmentation_utils.py:163]: Added extra dilation-erosion logic documentation [a72e4ee](https://github.com/Deci-AI/super-gradients/pull/239/commits/a72e4ee0b298ded8f842b8213d356330a86e6a0e)

**Comment by lkdci** [comment]
Addressed, added dilation-erosion documentation [a72e4ee](https://github.com/Deci-AI/super-gradients/pull/239/commits/a72e4ee0b298ded8f842b8213d356330a86e6a0e)

**Review by shaydeci — APPROVED** [review]
LGTM
